File: examples/gemma-ios-swift/Sources/GrantexGemmaExample/AuditLogger.swift
import Foundation
import CryptoKit

// MARK: - AuditLogger

/// Offline audit logger with hash chain and Ed25519 signing.
///
/// Every audit entry is:
/// 1. SHA-256 hashed together with the previous entry's hash (forming a chain)
/// 2. Signed with the Ed25519 private key from the consent bundle
/// 3. Appended to a JSONL file on disk
///
/// The hash chain makes the log tamper-evident: modifying any past entry
/// breaks the chain, and `verifyChain()` detects the exact break point.
///
/// Usage:
/// ```swift
/// let logger = AuditLogger()
/// try logger.initialize(auditKey: bundle.offlineAuditKey)
///
/// let entry = try logger.append(
///     action: "read_sensor",
///     grant: grant,
///     result: "success",
///     metadata: ["sensor_id": "temp_01"]
/// )
/// ```
class AuditLogger {

    /// The genesis hash — used as prev_hash for the first entry.
    static let genesisHash = String(repeating: "0", count: 64)

    /// Path to the JSONL audit log file.
    private let logPath: URL

    /// Current sequence number (auto-incremented).
    private var seq: Int = 0

    /// Hash of the previous entry (for chain linking).
    private var prevHash: String

    /// Ed25519 private key for signing entries.
    private var privateKey: Curve25519.Signing.PrivateKey?

    // MARK: - Init

    /// Create an audit logger.
    ///
    /// - Parameter logDirectory: Directory for the JSONL file. Defaults to
    ///   `~/.grantex/` on macOS or the app's documents directory on iOS.
    init(logDirectory: URL? = nil) {
        let dir: URL
        if let logDirectory {
            dir = logDirectory
        } else {
            #if os(macOS)
            dir = FileManager.default.homeDirectoryForCurrentUser
                .appendingPathComponent(".grantex")
            #else
            dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
                .appendingPathComponent("grantex")
            #endif
        }

        // Create directory if needed
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        self.logPath = dir.appendingPathComponent("audit.jsonl")
        self.prevHash = AuditLogger.genesisHash

        // Resume from existing log
        resumeFromDisk()
    }

    // MARK: - Initialize

    /// Initialize the logger with the Ed25519 signing key from the bundle.
    ///
    /// Must be called before `append()`.
    ///
    /// - Parameter auditKey: The Ed25519 key pair from the consent bundle.
    /// - Throws: If the private key cannot be parsed.
    func initialize(auditKey: OfflineAuditKey) throws {
        // Parse the Ed25519 private key from PEM
        let pem = auditKey.privateKey
        let b64 = pem
            .replacingOccurrences(of: "-----BEGIN PRIVATE KEY-----", with: "")
            .replacingOccurrences(of: "-----END PRIVATE KEY-----", with: "")
            .components(separatedBy: .whitespacesAndNewlines)
            .joined()

        guard let decoded = Data(base64Encoded: b64) else {
            throw GrantexError.bundleTampered("Invalid Ed25519 private key encoding")
        }

        // PKCS8 wrapping for Ed25519 has a prefix before the 32-byte key.
        // The raw key is the last 32 bytes.
        let keyBytes: Data
        if decoded.count > 32 {
            keyBytes = decoded.suffix(32)
        } else {
            keyBytes = decoded
        }

        privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: keyBytes)
    }

    // MARK: - Append

    /// Append a new signed, hash-chained audit entry.
    ///
    /// - Parameters:
    ///   - action: The action being audited (e.g., "read_sensor").
    ///   - grant: The verified grant that authorized this action.
    ///   - result: The outcome: "success", "denied", or "error".
    ///   - metadata: Optional key-value pairs with additional context.
    /// - Returns: The signed ``SignedAuditEntry``.
    /// - Throws: If signing fails.
    func append(
        action: String,
        grant: VerifiedGrant,
        result: String,
        metadata: [String: Any] = [:]
    ) throws -> SignedAuditEntry {
        guard let privateKey else {
            throw GrantexError.bundleTampered("Audit logger not initialized. Call initialize() first.")
        }

        seq += 1
        let timestamp = ISO8601DateFormatter().string(from: Date())

        // Build the entry dictionary for hashing
        let entryForHash: [String: Any] = [
            "seq": seq,
            "timestamp": timestamp,
            "action": action,
            "agent_did": grant.agentDID,
            "grant_id": grant.grantId,
            "scopes": grant.scopes,
            "result": result,
            "metadata": metadata,
            "prev_hash": prevHash,
        ]

        // Compute SHA-256 hash
        let hash = computeHash(entryForHash)

        // Sign the hash with Ed25519
        let hashData = hash.data(using: .utf8)!
        let signature = try privateKey.signature(for: hashData)
        let signatureB64 = signature.withUnsafeBytes { Data($0) }
            .base64URLEncodedString()

        // Build the full entry for storage
        let fullEntry: [String: Any] = [
            "seq": seq,
            "timestamp": timestamp,
            "action": action,
            "agent_did": grant.agentDID,
            "grant_id": grant.grantId,
            "scopes": grant.scopes,
            "result": result,
            "metadata": metadata,
            "prev_hash": prevHash,
            "hash": hash,
            "signature": signatureB64,
        ]

        // Serialize to JSON and append to the JSONL file
        let jsonData = try JSONSerialization.data(withJSONObject: fullEntry, options: [.sortedKeys])
        let jsonString = String(data: jsonData, encoding: .utf8)!

        let line = jsonString + "\n"
        if let fileHandle = try? FileHandle(forWritingTo: logPath) {
            fileHandle.seekToEndOfFile()
            fileHandle.write(line.data(using: .utf8)!)
            fileHandle.closeFile()
        } else {
            try line.write(to: logPath, atomically: false, encoding: .utf8)
        }

        // Update chain state
        prevHash = hash

        return SignedAuditEntry(
            seq: seq,
            timestamp: timestamp,
            action: action,
            agentDID: grant.agentDID,
            grantId: grant.grantId,
            scopes: grant.scopes,
            result: result,
            metadata: metadata,
            prevHash: entryForHash["prev_hash"] as! String,
            hash: hash,
            signature: signatureB64
        )
    }

    // MARK: - Read Entries

    /// Read all entries from the JSONL log file.
    func readEntries() -> [SignedAuditEntry] {
        guard let content = try? String(contentsOf: logPath, encoding: .utf8) else {
            return []
        }

        return content
            .split(separator: "\n")
            .filter { !$0.isEmpty }
            .compactMap { line -> SignedAuditEntry? in
                guard let data = line.data(using: .utf8),
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    return nil
                }

                return SignedAuditEntry(
                    seq: json["seq"] as? Int ?? 0,
                    timestamp: json["timestamp"] as? String ?? "",
                    action: json["action"] as? String ?? "",
                    agentDID: json["agent_did"] as? String ?? "",
                    grantId: json["grant_id"] as? String ?? "",
                    scopes: json["scopes"] as? [String] ?? [],
                    result: json["result"] as? String ?? "",
                    metadata: json["metadata"] as? [String: Any] ?? [:],
                    prevHash: json["prev_hash"] as? String ?? "",
                    hash: json["hash"] as? String ?? "",
                    signature: json["signature"] as? String ?? ""
                )
            }
    }

    // MARK: - Verify Chain

    /// Verify the integrity of the hash chain.
    ///
    /// - Returns: A tuple of (isValid, brokenAtIndex). If valid,
    ///   brokenAtIndex is nil.
    func verifyChain() -> (valid: Bool, brokenAt: Int?) {
        let entries = readEntries()
        guard !entries.isEmpty else {
            return (true, nil)
        }

        for (i, entry) in entries.enumerated() {
            // Recompute hash from content fields
            let entryForHash: [String: Any] = [
                "seq": entry.seq,
                "timestamp": entry.timestamp,
                "action": entry.action,
                "agent_did": entry.agentDID,
                "grant_id": entry.grantId,
                "scopes": entry.scopes,
                "result": entry.result,
                "metadata": entry.metadata,
                "prev_hash": entry.prevHash,
            ]

            let expectedHash = computeHash(entryForHash)
            if entry.hash != expectedHash {
                return (false, i)
            }

            // Check prev_hash linkage
            if i > 0 && entry.prevHash != entries[i - 1].hash {
                return (false, i)
            }
        }

        return (true, nil)
    }

    // MARK: - Internal

    /// Resume seq and prevHash from an existing log file.
    private func resumeFromDisk() {
        guard let content = try? String(contentsOf: logPath, encoding: .utf8) else {
            return
        }

        let lines = content.split(separator: "\n").filter { !$0.isEmpty }
        guard let lastLine = lines.last,
              let data = lastLine.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }

        seq = json["seq"] as? Int ?? 0
        prevHash = json["hash"] as? String ?? AuditLogger.genesisHash
    }

    /// Compute SHA-256 hash of an entry dictionary.
    ///
    /// Uses sorted keys for deterministic hashing that matches the
    /// Python and TypeScript implementations.
    private func computeHash(_ entry: [String: Any]) -> String {
        // Build canonical representation with sorted keys
        let canonical: [String: Any] = [
            "action": entry["action"] as Any,
            "agent_did": entry["agent_did"] as Any,
            "grant_id": entry["grant_id"] as Any,
            "metadata": entry["metadata"] as Any,
            "prev_hash": entry["prev_hash"] as Any,
            "result": entry["result"] as Any,
            "scopes": entry["scopes"] as Any,
            "seq": entry["seq"] as Any,
            "timestamp": entry["timestamp"] as Any,
        ]

        guard let jsonData = try? JSONSerialization.data(
            withJSONObject: canonical,
            options: [.sortedKeys]
        ) else {
            return ""
        }

        let hash = SHA256.hash(data: jsonData)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - SignedAuditEntry

/// A signed, hash-chained audit log entry.
struct SignedAuditEntry {
    let seq: Int
    let timestamp: String
    let action: String
    let agentDID: String
    let grantId: String
    let scopes: [String]
    let result: String
    let metadata: [String: Any]
    let prevHash: String
    let hash: String
    let signature: String
}

// MARK: - Data Extension

extension Data {
    /// Encode data as base64url (no padding).
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: examples/gemma-android-kotlin/app/src/main/java/dev/grantex/gemma/example/AuditLogger.kt
package dev.grantex.gemma.example

import android.content.Context
import android.util.Log
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileWriter
import java.security.MessageDigest
import java.time.Instant
import java.util.Base64

/**
 * Offline audit logger for Android with hash chain and Ed25519 signing.
 *
 * Every audit entry is:
 * 1. SHA-256 hashed with the previous entry's hash (forming an immutable chain)
 * 2. Signed with the Ed25519 private key from the consent bundle
 * 3. Appended to a JSONL file in the app's internal storage
 *
 * The hash chain ensures that if any historical entry is modified, the
 * chain breaks and the tampering is detectable. The Ed25519 signatures
 * prove that each entry was created by the authorized agent.
 *
 * Usage:
 * ```
 * val logger = AuditLogger(context)
 * logger.initialize(bundle.offlineAuditKey)
 *
 * val entry = logger.append(
 *     action = "read_sensor",
 *     agentDid = grant.agentDid,
 *     grantId = grant.grantId,
 *     scopes = grant.scopes,
 *     result = "success",
 *     metadata = mapOf("sensor_id" to "temp_01"),
 * )
 * ```
 */
class AuditLogger(private val context: Context) {

    companion object {
        private const val TAG = "AuditLogger"
        private const val LOG_DIR = "grantex"
        private const val LOG_FILE = "audit.jsonl"
        private const val GENESIS_HASH = "0000000000000000000000000000000000000000000000000000000000000000"
    }

    private var seq = 0
    private var prevHash = GENESIS_HASH
    private var privateKeyParams: Ed25519PrivateKeyParameters? = null
    private var logFile: File? = null

    /**
     * Initialize the audit logger with the Ed25519 signing key from the bundle.
     *
     * Must be called before [append]. Resumes the sequence number and
     * prev_hash from any existing log file on disk.
     *
     * @param auditKey The Ed25519 key pair from the consent bundle.
     */
    fun initialize(auditKey: OfflineAuditKey) {
        // Set up the log file in internal storage
        val dir = File(context.filesDir, LOG_DIR)
        dir.mkdirs()
        logFile = File(dir, LOG_FILE)

        // Parse the Ed25519 private key from PEM
        privateKeyParams = parseEd25519PrivateKey(auditKey.privateKey)

        // Resume state from existing log
        resumeFromDisk()

        Log.d(TAG, "Audit logger initialized (seq=$seq)")
    }

    /**
     * Append a new signed, hash-chained audit entry.
     *
     * @param action The action being audited (e.g., "read_sensor", "control_actuator").
     * @param agentDid The agent's DID from the verified grant.
     * @param grantId The grant ID from the verified grant.
     * @param scopes The grant's authorized scopes.
     * @param result The outcome: "success", "denied", or "error".
     * @param metadata Optional key-value pairs with additional context.
     * @return The signed [AuditEntry].
     */
    fun append(
        action: String,
        agentDid: String,
        grantId: String,
        scopes: List<String>,
        result: String,
        metadata: Map<String, Any> = emptyMap(),
    ): AuditEntry {
        requireNotNull(privateKeyParams) { "Call initialize() before append()" }
        requireNotNull(logFile) { "Call initialize() before append()" }

        seq++
        val timestamp = Instant.now().toString()

        // Build the entry (without hash and signature)
        val entryForHash = JSONObject().apply {
            put("seq", seq)
            put("timestamp", timestamp)
            put("action", action)
            put("agent_did", agentDid)
            put("grant_id", grantId)
            put("scopes", JSONArray(scopes))
            put("result", result)
            put("metadata", JSONObject(metadata))
            put("prev_hash", prevHash)
        }

        // Compute SHA-256 hash of the canonical JSON
        val hash = computeHash(entryForHash)

        // Sign the hash with Ed25519
        val signature = signData(hash)

        // Build the full entry JSON for storage
        val fullEntry = JSONObject().apply {
            put("seq", seq)
            put("timestamp", timestamp)
            put("action", action)
            put("agent_did", agentDid)
            put("grant_id", grantId)
            put("scopes", JSONArray(scopes))
            put("result", result)
            put("metadata", JSONObject(metadata))
            put("prev_hash", prevHash)
            put("hash", hash)
            put("signature", signature)
        }

        // Append to the JSONL file
        FileWriter(logFile!!, /* append = */ true).use { writer ->
            writer.write(fullEntry.toString() + "\n")
        }

        // Update chain state
        prevHash = hash

        val entry = AuditEntry(
            seq = seq,
            timestamp = timestamp,
            action = action,
            agentDid = agentDid,
            grantId = grantId,
            scopes = scopes,
            result = result,
            metadata = metadata,
            prevHash = entryForHash.getString("prev_hash"),
            hash = hash,
            signature = signature,
        )

        Log.d(TAG, "Audit #$seq: $action → $result (hash=${hash.take(12)}...)")
        return entry
    }

    /**
     * Read all audit entries from the log file.
     */
    fun readEntries(): List<AuditEntry> {
        val file = logFile ?: return emptyList()
        if (!file.exists()) return emptyList()

        return file.readLines()
            .filter { it.isNotBlank() }
            .map { line ->
                val json = JSONObject(line)
                AuditEntry(
                    seq = json.getInt("seq"),
                    timestamp = json.getString("timestamp"),
                    action = json.getString("action"),
                    agentDid = json.getString("agent_did"),
                    grantId = json.getString("grant_id"),
                    scopes = (0 until json.getJSONArray("scopes").length())
                        .map { json.getJSONArray("scopes").getString(it) },
                    result = json.getString("result"),
                    metadata = jsonObjectToMap(json.getJSONObject("metadata")),
                    prevHash = json.getString("prev_hash"),
                    hash = json.getString("hash"),
                    signature = json.getString("signature"),
                )
            }
    }

    /**
     * Verify the hash chain integrity of all stored entries.
     *
     * @return A [ChainVerificationResult] indicating validity.
     */
    fun verifyChain(): ChainVerificationResult {
        val entries = readEntries()
        if (entries.isEmpty()) {
            return ChainVerificationResult(valid = true, brokenAt = null, totalEntries = 0)
        }

        for (i in entries.indices) {
            val entry = entries[i]

            // Recompute the hash from the entry's content fields
            val entryForHash = JSONObject().apply {
                put("seq", entry.seq)
                put("timestamp", entry.timestamp)
                put("action", entry.action)
                put("agent_did", entry.agentDid)
                put("grant_id", entry.grantId)
                put("scopes", JSONArray(entry.scopes))
                put("result", entry.result)
                put("metadata", JSONObject(entry.metadata))
                put("prev_hash", entry.prevHash)
            }

            val expectedHash = computeHash(entryForHash)
            if (entry.hash != expectedHash) {
                return ChainVerificationResult(
                    valid = false,
                    brokenAt = i,
                    totalEntries = entries.size,
                )
            }

            // Check prev_hash linkage
            if (i > 0 && entry.prevHash != entries[i - 1].hash) {
                return ChainVerificationResult(
                    valid = false,
                    brokenAt = i,
                    totalEntries = entries.size,
                )
            }
        }

        return ChainVerificationResult(
            valid = true,
            brokenAt = null,
            totalEntries = entries.size,
        )
    }

    // ── Internal Helpers ────────────────────────────────────────────

    /**
     * Resume seq and prevHash from an existing log file on disk.
     */
    private fun resumeFromDisk() {
        val file = logFile ?: return
        if (!file.exists()) return

        val lines = file.readLines().filter { it.isNotBlank() }
        if (lines.isEmpty()) return

        try {
            val lastEntry = JSONObject(lines.last())
            seq = lastEntry.getInt("seq")
            prevHash = lastEntry.getString("hash")
        } catch (e: Exception) {
            Log.w(TAG, "Failed to resume from existing log, starting fresh", e)
        }
    }

    /**
     * Compute SHA-256 hash of a JSONObject's canonical string representation.
     *
     * Uses sorted keys and no whitespace for deterministic hashing that
     * matches the Python and TypeScript implementations.
     */
    private fun computeHash(entry: JSONObject): String {
        // Build a canonical representation with sorted keys
        val canonical = JSONObject().apply {
            put("action", entry.getString("action"))
            put("agent_did", entry.getString("agent_did"))
            put("grant_id", entry.getString("grant_id"))
            put("metadata", entry.getJSONObject("metadata"))
            put("prev_hash", entry.getString("prev_hash"))
            put("result", entry.getString("result"))
            put("scopes", entry.getJSONArray("scopes"))
            put("seq", entry.getInt("seq"))
            put("timestamp", entry.getString("timestamp"))
        }

        val payload = canonical.toString()
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(payload.toByteArray(Charsets.UTF_8))
        return hashBytes.joinToString("") { "%02x".format(it) }
    }

    /**
     * Sign data with the Ed25519 private key and return base64url signature.
     */
    private fun signData(data: String): String {
        val signer = Ed25519Signer()
        signer.init(true, privateKeyParams)
        val dataBytes = data.toByteArray(Charsets.UTF_8)
        signer.update(dataBytes, 0, dataBytes.size)
        val sigBytes = signer.generateSignature()
        return Base64.getUrlEncoder().withoutPadding().encodeToString(sigBytes)
    }

    /**
     * Parse an Ed25519 private key from PEM format.
     */
    private fun parseEd25519PrivateKey(pem: String): Ed25519PrivateKeyParameters {
        // Strip PEM headers and decode base64
        val b64 = pem
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replace("\\s".toRegex(), "")

        val decoded = Base64.getDecoder().decode(b64)

        // PKCS8 wrapping for Ed25519 has a 16-byte prefix before the 32-byte key
        // The raw key is the last 32 bytes of the decoded data
        val keyBytes = if (decoded.size > 32) {
            decoded.copyOfRange(decoded.size - 32, decoded.size)
        } else {
            decoded
        }

        return Ed25519PrivateKeyParameters(keyBytes, 0)
    }

    /**
     * Convert a JSONObject to a Map for the metadata field.
     */
    private fun jsonObjectToMap(json: JSONObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        json.keys().forEach { key ->
            map[key] = json.get(key)
        }
        return map
    }
}

// ── Data Classes ────────────────────────────────────────────────────

/**
 * A signed, hash-chained audit log entry.
 */
data class AuditEntry(
    val seq: Int,
    val timestamp: String,
    val action: String,
    val agentDid: String,
    val grantId: String,
    val scopes: List<String>,
    val result: String,
    val metadata: Map<String, Any>,
    val prevHash: String,
    val hash: String,
    val signature: String,
)

/**
 * Result of verifying the audit log hash chain.
 */
data class ChainVerificationResult(
    val valid: Boolean,
    val brokenAt: Int?,
    val totalEntries: Int,
)
